### PR TITLE
USD Skinned Mesh

### PR DESCRIPTION
**Pull request by AMZN-Gene**
- Converting TinyUSD skeleton and joints to Ai bones
- Skinned vertex weights

Simple test assets: 
1) [skintest.zip](https://github.com/user-attachments/files/17347336/skintest.zip)
2) https://developer.apple.com/augmented-reality/quick-look/models/drummertoy/toy_drummer_idle.usdz 
image
3) [simple_anim_posed.zip](https://github.com/user-attachments/files/17466816/simple_anim_posed.zip)
image

**Future Work**
1) Migrate to latest TinyUSDZ for duplicate skeleton fix. See https://github.com/assimp/assimp/pull/5849
2) Blender exports USD with joints (individual points), rather than bones (line segments) with _end bones like you see with FBX. As a result, Assimp Viewer skeleton may appear incomplete (missing the final bone). It might be worth investigating procedurally adding an _end bone.
3) Blender doesn't apply the USD's "rest" pose when reimporting a USD https://projects.blender.org/blender/blender/issues/116612
4) I believe TinyUSDZ sometimes loads incorrect bone weights, https://github.com/lighttransport/tinyusdz/issues/206

### CODE REVIEW

**Comment by AMZN-Gene** [comment]
@tellypresence you mentioned working on Animation in the comments for #5547.
What were you using to test? Does assimp-viewer work to play animations? I haven't loading animations yet, just the rig, so I'm also looking for a way to visualize the bones or bone weights as well.

File: code/AssetLib/USD/USDLoaderImplTinyusdz.cpp
Comment (AMZN-Gene) [reply code/AssetLib/USD/USDLoaderImplTinyusdz.cpp:73]: Thanks! That snuck if when I was debugging without debug config. Will remove :)

**Comment by AMZN-Gene** [comment]
Update:
Skeleton (red lines) now appear in proper model space, and is standing upright, but the 1st bone appears to be missing. Notice in Blender the first bone is standing straight up, while in Assimp Viewer the first bone is angled.
image

**Comment by turol** [comment]
Could you add some unit tests?

**Comment by AMZN-Gene** [comment]
> Could you add some unit tests?

Yes! I'll look for some existing examples of skinning tests with other file types. If you have a place to look, that'd be greatly appreciated (my 1st assimp PR)

**Comment by tellypresence** [comment]
@AMZN-Gene Raised PR #5849 which updates tinyusdz git commit and patch file

**Comment by tellypresence** [comment]
@AMZN-Gene
A few suggestions to make sure build won't break in environments treating warnings as errors:
- Unused variable, `code/AssetLib/USD/USDLoaderImplTinyusdz.cpp` line 306:
   - `const int numJointIndices = jointIndices.size();`

**Comment by tellypresence** [comment]
Confirm that with this PR combined with PR #5849 Apple robot model indicates it has 240 bones:
apple_robot_w_bones

and Apple toy drummer model has 410 bones:
apple_toy_drummer_410_bones

File: code/AssetLib/USD/USDLoaderImplTinyusdz.cpp
Comment (tellypresence) [inline code/AssetLib/USD/USDLoaderImplTinyusdz.cpp:306]: Unused variable `numJointIndices `, will break build if warnings-as-errors enabled

**Review by tellypresence — CHANGES_REQUESTED** [review]

File: code/AssetLib/USD/USDLoaderImplTinyusdz.cpp
Comment (AMZN-Gene) [reply code/AssetLib/USD/USDLoaderImplTinyusdz.cpp]: Thank you! Fixed!

Comment (AMZN-Gene) [reply code/AssetLib/USD/USDLoaderImplTinyusdz.cpp]: Fixed! Thank you!

**Comment by AMZN-Gene** [comment]
> @AMZN-Gene Raised PR #5849 which updates tinyusdz git commit and patch file -- please verify that it works in your environment, thank you

Works! I only have to iterate one tiny USD skeleton, instead of one per sub-mesh, in order to discover bones. Thank you :)

**Comment by AMZN-Gene** [comment]
> Could you add some unit tests?

Unit tests added :D

**Comment by tellypresence** [comment]
> > Could you add some unit tests?
> 
> Unit tests added :D

Ah, don't think unit tests will work for USD as things are currently configured.  `tinyusdz` is opt-in (default is off, so the `tinyusdz` repo isn't even cloned) so CI won't have access to the code, which means USD models can't be loaded...so USD unit tests will fail

**Comment by AMZN-Gene** [comment]
> > > Could you add some unit tests?
> > 
> > 
> > Unit tests added :D
> 
> Ah, don't think unit tests will work for USD as things are currently configured. `tinyusdz` is opt-in (default is off, so the `tinyusdz` repo isn't even cloned) so CI won't have access to the code, which means USD models can't be loaded...so USD unit tests will fail
> 
> The reason `tinyusdz` had to be turned off by default, is because it caused CI checks to fail due to compiler warnings, details in [comment](https://github.com/assimp/assimp/pull/5628#issuecomment-2224585088) on
> 
> * [[USD] Integrate "tinyusdz" project #5628](https://github.com/assimp/assimp/pull/5628)
> 
> To enable `tinyusdz` by default, exceptions for the offending warnings would need to be added to assimp's CI rules

I've updated the cmake to only add USD unit tests if USD_IMPORT flag is enabled. I think this is the best we can do to allow for quick testing locally, until the TinyUSDZ repo fixes the compiler warnings.

**Review by moudgils — APPROVED** [review]
I am not a maintainer of this codebase but am looking forward to these change for O3DE. The code looks good to me.

**Comment by AMZN-Gene** [comment]
Is removing the [6 new issues](https://sonarcloud.io/project/issues?id=assimp_assimp&pullRequest=5812&issueStatuses=OPEN,CONFIRMED&sinceLeakPeriod=true) a requirement before getting merged?

**Comment by AMZN-Gene** [comment]
Re-pinging @tellypresence and @kimkulling for feedback

**Comment by kimkulling** [comment]
No! I just need time next to my day job and my family to do all the reviews.

**Comment by tellypresence** [comment]
@AMZN-Gene I just built with the latest assimp commit, which contains this PR plus the latest tinyusdz PR, and still get 410 bones for the drummer and 240 bones for the robot -- please build with the latest and check your results, thank you

**Comment by AMZN-Gene** [comment]
> @AMZN-Gene I just built with the latest assimp commit, which contains this PR plus the latest tinyusdz PR, and still get 410 bones for the drummer and 240 bones for the robot -- please build with the latest and check your results, thank you

Correct. The drummer really only has 10 bones, but because assimp stores mBones per mesh, and there are 41 meshes, each one is storing the skeleton for itself. There could be an optimization perhaps where we remove bones from a mesh if the bone doesn't have an any weighted influence on its vertex. We'd also have to be careful to keep the bone's ancestors, even if the ancestor->mWeights is empty. Thoughts?

**Pull request merged**
